Installing Slumber causes error: unresolved imports `signal_hook::consts::SIGHUP`, `signal_hook::consts::SIGQUIT`, `signal_hook::iterator`

**Did you [search](https://github.com/LucasPickering/slumber/issues) for existing issues already?**
Yes

**Describe the bug**
I am running on windows and opened the terminal. In git bash, I ran `cargo install slumber`. 

The app downloaded and started to compile. On the last step, I got this error:


```
 Compiling hyper v0.14.28
   Compiling hyper-rustls v0.24.2
   Compiling reqwest v0.11.27
   Compiling slumber v1.0.0
error[E0432]: unresolved imports `signal_hook::consts::SIGHUP`, `signal_hook::consts::SIGQUIT`, `signal_hook::iterator`
  --> C:\Users\22\.cargo\registry\src\index.crates.io-6f17d22bba15001f\slumber-1.0.0\src\tui.rs:29:14
   |
29 |     consts::{SIGHUP, SIGINT, SIGQUIT, SIGTERM},
   |              ^^^^^^          ^^^^^^^ no `SIGQUIT` in `consts`
   |              |
   |              no `SIGHUP` in `consts`
30 |     iterator::Signals,
   |     ^^^^^^^^ could not find `iterator` in `signal_hook`

For more information about this error, try `rustc --explain E0432`.
error: could not compile `slumber` (bin "slumber") due to 1 previous error
error: failed to compile `slumber v1.0.0`, intermediate artifacts can be found at `C:\Users\22~1\AppData\Local\Temp\cargo-installJnLI65`.
To reuse those artifacts with a future compilation, set the environment variable `CARGO_TARGET_DIR` to that path.
```

**To Reproduce**
_Steps to reproduce the behavior_

1.

**Expected behavior**
_A clear and concise description of what you expected to happen_

**Screenshots**
_If applicable, add screenshots to help explain your problem_

**Version (please complete the following information):**
- Cargo 1.77.0
- rustc 1.77.0

- OS:
- Terminal: GitBash
- Slumber Version:

**Additional context**
Add any other context about the problem here.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,16 @@
 # Changelog
 
+## [Unreleased] - ReleaseDate
+
+### Added
+
+- Add two new build targets to releases: `x86_64-pc-windows-msvc` and `x86_64-unknown-linux-musl`
+
+### Fixed
+
+- Fix build on Windows ([#180](https://github.com/LucasPickering/slumber/issues/180))
+  - I can't guarantee it _works_ on Windows since I don't have a machine to test on, but it at least compiles now
+
 ## [1.0.0] - 2024-04-25
 
 ### Breaking

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -42,10 +42,9 @@ serde = {version = "^1.0.188", features = ["derive"]}
 serde_json = {version = "^1.0.107", default-features = false}
 serde_json_path = "^0.6.3"
 serde_yaml = {version = "^0.9.25", default-features = false}
-signal-hook = "^0.3.17"
 strum = {version = "^0.26.0", default-features = false, features = ["derive"]}
 thiserror = "^1.0.48"
-tokio = {version = "^1.32.0", default-features = false, features = ["fs", "macros", "process", "rt", "rt-multi-thread"]}
+tokio = {version = "^1.32.0", default-features = false, features = ["fs", "macros", "process", "rt", "rt-multi-thread", "signal"]}
 tracing = "^0.1.37"
 tracing-subscriber = {version = "^0.3.17", default-features = false, features = ["ansi", "env-filter", "fmt", "registry"]}
 url = {version = "^2.5.0", features = ["serde"]}

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -75,11 +75,17 @@ cargo-dist-version = "0.13.3"
 # CI backends to support
 ci = ["github"]
 # The installers to generate for each app
-installers = ["shell", "homebrew"]
+installers = ["shell", "homebrew", "powershell"]
 # A GitHub repo to push Homebrew formulas to
 tap = "LucasPickering/homebrew-tap"
 # Target platforms to build apps for (Rust target-triple syntax)
-targets = ["aarch64-apple-darwin", "x86_64-apple-darwin", "x86_64-unknown-linux-gnu", "x86_64-unknown-linux-musl"]
+targets = [
+  "aarch64-apple-darwin",
+  "x86_64-apple-darwin",
+  "x86_64-unknown-linux-gnu",
+  "x86_64-unknown-linux-musl",
+  "x86_64-pc-windows-msvc",
+]
 # Publish jobs to run in CI
 publish-jobs = ["homebrew"]
 # Publish jobs to run in CI

--- a/src/tui.rs
+++ b/src/tui.rs
@@ -1,6 +1,7 @@
 mod context;
 pub mod input;
 mod message;
+mod signal;
 mod view;
 
 use crate::{
@@ -13,6 +14,7 @@ use crate::{
         context::TuiContext,
         input::Action,
         message::{Message, MessageSender, RequestConfig},
+        signal::signals,
         view::{ModalPriority, PreviewPrompter, RequestState, View},
     },
     util::Replaceable,
@@ -25,10 +27,6 @@ use crossterm::{
 use futures::Future;
 use notify::{event::ModifyKind, RecursiveMode, Watcher};
 use ratatui::{prelude::CrosstermBackend, Terminal};
-use signal_hook::{
-    consts::{SIGHUP, SIGINT, SIGQUIT, SIGTERM},
-    iterator::Signals,
-};
 use std::{
     io::{self, Stdout},
     ops::Deref,
@@ -127,10 +125,7 @@ impl Tui {
     /// the struct definition for a description of the different phases of the
     /// run loop.
     fn run(mut self) -> anyhow::Result<()> {
-        // Listen for signals to stop the program
-        let mut quit_signals = Signals::new([SIGHUP, SIGINT, SIGTERM, SIGQUIT])
-            .context("Error creating signal handler")?;
-
+        self.listen_for_signals();
         // Hang onto this because it stops running when dropped
         let _watcher = self.watch_collection()?;
 
@@ -174,11 +169,6 @@ impl Tui {
 
             // ===== Draw Phase =====
             self.terminal.draw(|f| self.view.draw(f))?;
-
-            // ===== Signal Phase =====
-            if quit_signals.pending().next().is_some() {
-                self.quit();
-            }
         }
 
         Ok(())
@@ -293,6 +283,16 @@ impl Tui {
             }
         }
         Ok(())
+    }
+
+    /// Spawn a task to listen in the backgrouns for quit signals
+    fn listen_for_signals(&self) {
+        let messages = self.messages_tx.clone();
+        self.spawn(async move {
+            signals().await?;
+            messages.send(Message::Quit);
+            Ok(())
+        });
     }
 
     /// Spawn a watcher to automatically reload the collection when the file

--- a/src/tui/signal.rs
+++ b/src/tui/signal.rs
@@ -1,0 +1,50 @@
+//! Utilities for handling signals to the process
+
+use futures::{future, FutureExt};
+use tracing::info;
+
+/// Listen for any exit signals, and return `Ok(())` when any signal is
+/// received. This can only fail during initialization.
+#[cfg(unix)]
+pub async fn signals() -> anyhow::Result<()> {
+    use anyhow::Context;
+    use itertools::Itertools;
+    use tokio::signal::unix::{signal, Signal, SignalKind};
+
+    let signals: Vec<(Signal, SignalKind)> = [
+        SignalKind::interrupt(),
+        SignalKind::hangup(),
+        SignalKind::terminate(),
+        SignalKind::quit(),
+    ]
+    .into_iter()
+    .map::<anyhow::Result<_>, _>(|kind| {
+        let signal = signal(kind).with_context(|| {
+            format!("Error initializing listener for signal `{kind:?}`")
+        })?;
+        Ok((signal, kind))
+    })
+    .try_collect()?;
+    let futures = signals
+        .into_iter()
+        .map(|(mut signal, kind)| async move {
+            signal.recv().await;
+            info!(?kind, "Received signal");
+        })
+        .map(FutureExt::boxed);
+    future::select_all(futures).await;
+    Ok(())
+}
+
+/// Listen for any exit signals, and return `Ok(())` when any signal is
+/// received. This can only fail during initialization.
+#[cfg(windows)]
+pub async fn signals() -> anyhow::Result<()> {
+    use tokio::signal::windows::{ctrl_break, ctrl_c, ctrl_close};
+
+    let (mut s1, mut s2, mut s3) = (ctrl_c()?, ctrl_break()?, ctrl_close()?);
+    let futures = vec![s1.recv().boxed(), s2.recv().boxed(), s3.recv().boxed()];
+    future::select_all(futures).await;
+    info!("Received exit signal");
+    Ok(())
+}

